nautobot/dcim/forms.py

```diff
@@ -2064,6 +2064,35 @@ def __init__(self, *args, **kwargs):
         if position:
             self.fields["position"].widget.choices = [(position, f"U{position}")]
 
+    def clean(self):
+        super().clean()
+
+        device_type = self.cleaned_data["device_type"]
+        software_version = self.cleaned_data["software_version"]
+        software_image_files = self.cleaned_data["software_image_files"]
+
+        # If software version is specified and any software image file is specified, validate that
+        # each of the software image files belongs to the specified software version's software image files
+        # that match the device's device type or is a default image
```

Comment:
Why do we need to duplicate the logic from the model clean here in the form? Is the model clean insufficient to provide the desired UI behavior?

Reply:
Yeah, I think we have to validate the image file in the `form.clean()` before it makes to `model.clean()` in the editing case.
Because `self.software_image_files` attribute is not updated yet with the new data from the form in `model.clean()`. It will still have outdated but valid image files and it will pass the validation logic in `model.clean()`. The device will have the latest but invalid image files as a result.

Reply:
Ah, because `software_image_files` is a many-to-many, I'm guessing? That makes sense.